### PR TITLE
feat: Add chat rate limit

### PR DESCRIPTION
**Pull request by leehuwuj**

### CODE REVIEW

File: src/ragapp/backend/models/orm/request.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+from typing import Optional
+
+from sqlmodel import Field, Index, SQLModel
+
+
+class UserChatRequest(SQLModel, table=True):
```
Comment (marcusschiesser) [inline src/ragapp/backend/models/orm/request.py:6]: why not storing a `User` model that has a `ratelimit` object (in the ORM sense) like this:
```
"ratelimit": {
  "time_frame",
  "request_count",
}
```
I am not sure if `time_frame` is a better name for `window_frame` - where did you get that term from?

File: src/ragapp/backend/tests/test_request_limit_middleware.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
+import datetime
```
Comment (marcusschiesser) [inline src/ragapp/backend/tests/test_request_limit_middleware.py:1]: yeah, we need more tests - we should run them in github actions

File: src/ragapp/backend/models/orm/request.py (hunk 1 above)
Comment (leehuwuj) [reply src/ragapp/backend/models/orm/request.py:6]: Ah, i just inspired the name from the sliding window approach (which should have window start and stop pointer and aggregate the total request in the windows), but our one is much simpler and i think it's enough for our case. But your suggestion `time_frame` sounds easier to understand, i'll update it.
fyi, it's better to use this: https://github.com/laurents/slowapi but it's not using db or file for persistant, so i just came up with the current implementation.

Comment (leehuwuj) [reply src/ragapp/backend/models/orm/request.py:6]: > why not storing a User model that has a ratelimit object (in the ORM sense) like this:
  "ratelimit": {
    "time_frame",
    "request_count",
  }
I'm not sure about how we are going to add more field to the User table so using a dedicated table for the request is not a bad idea. As it's frequently read and update per request so normalize it into the current table also easier to indexing and reduce the db performance when fetching the data.

File: .changeset/quick-maps-tan.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+"ragbox": patch
+---
+
+Add rate limit for chat api
```
Comment (marcusschiesser) [inline .changeset/quick-maps-tan.md:5]:
```suggestion
Add rate limiting (number of chat messages per user) to chat API
```

File: src/manager/app/models/ragapp.py [hunk 4]
```diff
@@ -35,6 +35,7 @@ def get_default_app_environment(app_name: str) -> Dict[str, str]:
         "MODEL": "gpt-4o-mini",
         "EMBEDDING_MODEL": "text-embedding-3-small",
         "EMBEDDING_DIM": "1024",
+        "CHAT_REQUEST_LIMIT_THRESHOLD": "20",  # 20 chat requests per day
```
Comment (marcusschiesser) [inline src/manager/app/models/ragapp.py:38]: how about we pass this via manager ENV `CHAT_REQUEST_LIMIT_THRESHOLD`?

File: src/ragapp/backend/middleware.py
Comment (marcusschiesser) [inline src/ragapp/backend/middleware.py:1]: how about moving the middlewares to a own folder?
`src/ragapp/backend/middlewares/rate_limiting.py`

File: src/ragapp/backend/middleware.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        )
+    except jwt.InvalidTokenError:
+        raise HTTPException(
+            status_code=status.HTTP_401_UNAUTHORIZED, detail="Invalid authentication"
+        )
+
+
+def _decode_jwt(token: str):
+    try:
+        # Remove Barear prefix
```
Comment (marcusschiesser) [inline src/ragapp/backend/middleware.py:65]:
```suggestion
        # Remove Bearer prefix
```

File: src/ragapp/backend/middleware.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+JWT_COOKIE_NAME = "Authorization"  # The name of the cookie that stores the JWT token
+JWT_USER_ID_CLAIM = "preferred_username"  # The claim in the JWT token that stores the user ID or user name
+CHAT_REQUEST_LIMIT_THRESHOLD = int(os.environ.get("CHAT_REQUEST_LIMIT_THRESHOLD", 5))
+
+
+async def request_limit_middleware(request: Request) -> Response:
+    time_frame = _get_time_frame()
+    user_id = _extract_user_id_from_request(request)
+    db: Session = next(DB.get_session())
+    request_count = get_user_chat_request_count(db, user_id, time_frame)
```
Comment (marcusschiesser) [inline src/ragapp/backend/middleware.py:25]: how about using OO?
```suggestion
    service = UserChatService()
    request_count = service.get_request_count(user_id, time_frame)
```
`UserChatService` can then look up the db session

Comment (marcusschiesser) [reply src/ragapp/backend/middleware.py:25]: we might then also move this service to a `services` folder

File: src/ragapp/backend/models/orm/request.py (hunk 1 above)
Comment (marcusschiesser) [reply src/ragapp/backend/models/orm/request.py]: >dedicated table for the request is not a bad idea
agree

how about using two model classes then?
`User` and `UserChatRequest ` where as the later uses the `user_id` as foreign key?
`User` can then store later the usual stuff like name and email

File: src/manager/app/models/ragapp.py (hunk 4 above)
Comment (leehuwuj) [reply src/manager/app/models/ragapp.py:38]: Ah, great idea. We can then easily change limit in deployment without update the code

File: src/ragapp/backend/middleware.py (hunk 6 above)
Comment (leehuwuj) [reply src/ragapp/backend/middleware.py]: I moved to `UserChatService` but haven't removed the db params yet. The reason is i think we will soon need to add management api to clear/refresh the rate limit - and the db session is a dependency for the api request which is injected automatically.

File: deployments/multiple-ragapps/docker-compose.yml [hunk 7]
```diff
@@ -66,6 +66,7 @@ services:
     environment:
       - BASE_URL=/manager
       - RAGAPP_IMAGE=${RAGAPP_IMAGE:-ragapp/ragapp:latest}
+      - CHAT_REQUEST_LIMIT_THRESHOLD=5
```
Comment (marcusschiesser) [inline deployments/multiple-ragapps/docker-compose.yml:69]: describe feature in README.md

File: src/ragapp/backend/middleware.py (hunk 6 above)
Comment (marcusschiesser) [reply src/ragapp/backend/middleware.py]: how about adding the session to the constructor of the `UserChatService`?
if it's not passed `DB.get_session()` is used per default - then you can still pass the session if needed and we would have here:
```
service = UserChatService()
request_count = service.get_request_count(user_id, time_frame)
```

Comment (leehuwuj) [reply src/ragapp/backend/middleware.py]: Ok, so i just moved the db session to the UserChatService. I'll add the session parameter once we implement new feature that need to reuse the available db session :D

**Pull request merged**